crates/core/src/factories/stub_transport.rs

```diff
@@ -0,0 +1,341 @@
+                        data_send.reason = Some(format!("{err:?}"));
+                    }
+                    let _ = res.send(Err(err));
+                } else {
+                    let _ = res.send(Ok(()));
+                }
+            }
+            Cmd::Disconnect(url, payload, res) => {
+                if let Some(mut data_send) = con_pool.remove(&url) {
+                    if let Some((reason, payload)) = payload {
+                        data_send.reason = Some(reason);
+                        let _ = data_send.send.send((payload, res));
+                    }
+                }
+            }
+            Cmd::Send(url, data, res) => {
+                if let Some(send) = get_stat().connect(
+                    &cmd_send,
+                    &mut con_pool,
+                    &url,
+                    &this_url,
+                ) {
+                    let _ = send.send((data, res));
+                }
+            }
+        }
+    }
+}
+
+struct Listener {
+    id: u64,
+    url: Url,
+    recv: ConRecv,
+}
+
+impl std::fmt::Debug for Listener {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Listener").field("url", &self.url).finish()
+    }
+}
+
+impl Drop for Listener {
+    fn drop(&mut self) {
+        get_stat().remove(self.id);
+    }
+}
+
+impl Listener {
+    pub fn url(&self) -> Url {
+        self.url.clone()
+    }
+}
+
+struct Stat {
+    con_map: Mutex<HashMap<u64, ConSend>>,
+}
+
+impl Stat {
+    fn new() -> Self {
+        Self {
+            con_map: Mutex::new(HashMap::new()),
+        }
+    }
+
+    fn listen(&self) -> Listener {
+        use std::sync::atomic::*;
+        static ID: AtomicU64 = AtomicU64::new(1);
+        let id = ID.fetch_add(1, Ordering::Relaxed);
+        let url = Url::from_str(format!("ws://stub.tx:42/{id}")).unwrap();
+        let (send, recv) = tokio::sync::mpsc::unbounded_channel();
+        self.con_map.lock().unwrap().insert(id, send);
+        Listener { id, url, recv }
+    }
+
+    fn remove(&self, id: u64) {
+        self.con_map.lock().unwrap().remove(&id);
+    }
+
+    fn connect(
+        &self,
+        cmd_send: &CmdSend,
+        map: &mut HashMap<Url, DropSend>,
+        to_peer: &Url,
+        from_peer: &Url,
+    ) -> Option<DataSend> {
+        if let Some(send) = map.get(to_peer) {
+            return Some(send.send.clone());
+        }
+
+        let id: u64 = match to_peer.peer_id() {
+            None => return None,
+            Some(id) => match id.parse() {
+                Err(_) => return None,
+                Ok(id) => id,
+            },
+        };
+
+        let send = match self.con_map.lock().unwrap().get(&id) {
+            None => return None,
+            Some(send) => send.clone(),
+        };
+
+        let (ds1, dr1) = tokio::sync::mpsc::unbounded_channel();
+        let (ds2, dr2) = tokio::sync::mpsc::unbounded_channel();
```

Comment:
Just checking the two channels here? The `listen` set up a channel for the Stat to message us when somebody sends it a message for us. Then we need a way to send message to the stat but also to respond?

Reply:
Network connections are bi-directional. Channels are uni-directional. We need two of them so our peer and us can both send and receive messages.

Reply:
Cool, yes that makes sense. Thank you